### PR TITLE
Implement Resolver mechanism within ConductR CLI

### PR DESCRIPTION
**Pull request by fsat**
Part of the task required to implement #72 
- [x] Encapsulate existing file and http url resolution into uri-based resolver
- [x] Implement caching mechanism
- [x] Implement bintray-based resolver supporting bundle short-hand expression (as per #72)

### CODE REVIEW

**Comment by fsat** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;i7nnHd2j&#x2F;138-shorthand-for-bundle-expressions"> Shorthand for bundle expressions</a>

File: conductr_cli/conduct.py [hunk 1]
```diff
@@ -8,6 +8,9 @@
 
 DEFAULT_PORT = os.getenv('CONDUCTR_PORT', '9005')
 DEFAULT_API_VERSION = os.getenv('CONDUCTR_API_VERSION', '1')
+DEFAULT_CLI_SETTINGS_DIR = os.getenv('CONDUCTR_CLI_SETTINGS_DIR', '{}/.conductr'.format(os.getenv('HOME')))
```
Comment (huntc) [inline conductr_cli/conduct.py]: Will that work on Windoze?

File: conductr_cli/resolvers/bintray_resolver.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            print('Bintray credentials loaded from {}'.format(BINTRAY_CREDENTIAL_FILE_PATH))
+            return username, password
+
+
+def bintray_download_url(bintray_username, bintray_password, org, repo, package_name, compatibility_version, digest):
+    if compatibility_version is None and digest is None:
+        # Get latest version
+        package_endpoint = '{}/packages/{}/{}/{}'.format(BINTRAY_API_BASE_URL, org, repo, package_name)
+        package = get_json(package_endpoint, bintray_username, bintray_password)
+        latest_version = package['latest_version']
```
Comment (huntc) [inline conductr_cli/resolvers/bintray_resolver.py:72]: Unfortunately I've seen this field set to null having published just one version. I don't think we should rely on it. Instead, we should always get the package and iterate through the `latest-v` property names to determine which one is the latest.

File: conductr_cli/resolvers/bintray_resolver.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            raise BintrayResolutionError(
+                'Unable to find version for owner={} repo={} package={} version={}'.format(
+                    org, repo, package_name, bintray_version))
+        else:
+            path = matching_files[0]['path']
+            download_url = '{}/{}/{}/{}'.format(BINTRAY_DOWNLOAD_BASE_URL, org, repo, path)
+            return download_url
+
+
+def get_version_timestamp(bintray_username, bintray_password, org, repo, package_name, version):
```
Comment (huntc) [inline conductr_cli/resolvers/bintray_resolver.py]: I don't think that this function is used anymore.

File: conductr_cli/resolvers/bintray_resolver.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        response = requests.get(uri, auth=(username, password))
+    else:
+        response = requests.get(uri)
+    response.raise_for_status()
+    return json.loads(response.text)
+
+
+def log(message, org, repo, package_name, compatibility_version, digest):
+    if compatibility_version is None and digest is None:
+        print('{} org={} repo={} bundle={} version={}'.format(message, org, repo, package_name, 'latest'))
```
Comment (huntc) [inline conductr_cli/resolvers/bintray_resolver.py]: We should use the shorthand convention here i.e.

{org}/{repo}/{package_name}:latest

...same for the others below.

File: conductr_cli/resolvers/bintray_resolver.py (hunk 2 above)
Comment (fsat) [inline conductr_cli/resolvers/bintray_resolver.py:72]: @huntc - I've implemented the change to infer latest version from list of attribute names that starts with `latest-v` below

**Pull request merged**